bundle/src/main/java/com/adobe/acs/commons/indesign/dynamicdeckdynamo/services/impl/XMLGeneratorServiceImpl.java

```diff
@@ -0,0 +1,301 @@
+import org.apache.sling.api.resource.ValueMap;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+import org.xml.sax.SAXException;
+
+import javax.xml.XMLConstants;
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.transform.Result;
+import javax.xml.transform.TransformerException;
+import javax.xml.transform.TransformerFactory;
+import javax.xml.transform.dom.DOMSource;
+import javax.xml.transform.stream.StreamResult;
+import java.io.*;
+import java.nio.file.Files;
+import java.nio.file.StandardCopyOption;
+import java.util.*;
+
+/**
+ * This class is used to parse Annotated Template XML and generate Indd XML
+ */
+@Component(service = XMLGeneratorService.class)
+public class XMLGeneratorServiceImpl implements XMLGeneratorService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(XMLGeneratorServiceImpl.class);
+
+    /**
+     * This Method Parses an Annotated Template XML and Creates an INDD XML out of the Template. The INDD XML is sent to
+     * INDD Server for generation of decks.
+     */
+    @Override
+    public String generateInddXML(InputStream xmlInputStream,
+                                  List<XMLResourceIterator> assetItrList, Resource masterResource, Resource deckResource,
+                                  ResourceResolver resourceResolver, List<String> imageList) throws DynamicDeckDynamoException {
+
+        InputStream resultInputStream;
+        Asset processXmlAsset = null;
+        final String xmlName = "processedXml.xml";
+
+        if (null == masterResource) {
+            masterResource = deckResource;
+        }
+
+        File targetFile = new File("targetFile-" + Calendar.getInstance().getTimeInMillis() + DynamicDeckDynamoConstants.DASH
+                + StringUtils.substringBeforeLast(xmlName, DynamicDeckDynamoConstants.DOT) + ".tmp");
+
+        try (ByteArrayOutputStream outputStream = new ByteArrayOutputStream()) {
+
+            Files.copy(xmlInputStream, targetFile.toPath(), StandardCopyOption.REPLACE_EXISTING);
+
+            DocumentBuilder dBuilder = DocumentBuilderFactory.newInstance().newDocumentBuilder();
+            Document doc = dBuilder.parse(targetFile);
+
+            if (doc.hasChildNodes()) {
+                populateDocNodes(doc, assetItrList, masterResource, imageList);
+            }
+
+            Result outputTarget = new StreamResult(outputStream);
+            TransformerFactory factory = TransformerFactory.newInstance();
+            factory.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
+            factory.newTransformer().transform(new DOMSource(doc), outputTarget);
+            resultInputStream = new ByteArrayInputStream(outputStream.toByteArray());
```

Comment:
you are creating a temporary Array for the complete outputstream in memory. And depending on the size of the file this can be quite intense.The problem is that this amount of memory needs to be available as single block; forcing a JVM to provide 300 MB block of continuous heap is causing a lot of garbage collection. 

Please refrain from doing so and use a different approach to get that outputstream into an inputstream.

Reply:
Not sure, about the alternative solution for this.

Reply:
As mentioned above, you can use a temporary file for the *output* whereas it is unnecessary for the *input* (the input can just be the passed `xmlInputStream`)

Reply:
Actually "xmlInputStream" is a template XML which we are processing & populating it with value/metadata/image-path/etc in `populateDocNodes(doc, assetItrList, masterResource, imageList);`. So, in `Document doc` we have processed output xml which we need to create asset in aem.

Reply:
Here's my understanding of the current flow.

`xmlInputStream` -> temporary file -> DOM Document -> (processing) -> `byte[]` -> Asset

This has two problems:
1. The temporary file used to generate the DOM Document object is unnecessary (and not cleaned up)
2. The entire XML document is stored in memory twice -- once as a DOM object and once as a byte array.

The better flow IMO would be:

`xmlInputStream` -> DOM Document -> (processing) -> temporary file -> Asset

Better still would be to be to do the processing in a streaming fashion using StAX as this would ensure minimal memory usage, but I suspect that this is a non-trivial change to the processing logic. But if you did that it would look like

`xmlInputStream` -> XML Event Stream -> (processing) -> XML Event Stream -> temporary file -> Asset

Reply:
Thanks @justinedelson, crystal clear by the (->) flow explanation.
I have converted the execution flow to
xmlInputStream -> DOM Document -> (processing which updates the DOM Document) -> temporary file -> Asset